eth/downloader/statesync.go

```diff
@@ -0,0 +1,245 @@
+	s := &stateSync{
+		d:        d,
+		db:       db,
+		sched:    state.NewStateSync(root, db),
+		fetching: make(map[string]*stateReq),
+		timeout:  make(chan *stateReq),
+		quit:     make(chan struct{}),
+	}
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.err = s.run(d.stateCh)
+	}()
+	return s
+}
+
+// cancel aborts a running state sync. It can be called multiple times.
+func (s *stateSync) cancel() {
+	s.quitOnce.Do(func() {
+		close(s.quit)
+		s.wait()
+	})
+}
+
+// wait blocks until the sync is done or canceled.
+func (s *stateSync) wait() error {
+	s.wg.Wait()
+	return s.err
+}
+
+func (s *stateSync) run(deliver <-chan dataPack) error {
+	defer func() {
+		for _, req := range s.fetching {
+			s.requestDone(req)
+		}
+	}()
+
+	newPeer := make(chan *peer, 200)
+	peerSub := s.d.peers.SubscribeNewPeers(newPeer)
+	defer peerSub.Unsubscribe()
+
+	for {
+		if s.sched.Pending() == 0 {
+			return nil
+		}
+		s.assignTasks()
+
+		select {
+		case packI := <-deliver:
+			pack := packI.(*statePack)
+			req := s.fetching[pack.peerId]
+			if req == nil {
+				log.Debug("Unrequested node data", "peer", pack.peerId, "len", len(pack.states))
+				continue
+			}
+			s.requestDone(req)
+			req.peer.SetNodeDataIdle(pack.Items())
+			if pack.Items() == 0 {
+				log.Trace("Empty node data pack", "peer", pack.peerId)
+				for _, item := range req.items {
+					req.peer.MarkLacking(item)
+				}
+				continue
+			}
+			procStart := time.Now()
+			n, err := s.process(req, pack)
+			if err != nil {
+				log.Warn("Node data write error", "err", err)
+			}
+			s.updateStats(n, time.Since(procStart))
+
+		case req := <-s.timeout:
+			if s.fetching[req.peer.id] != req {
+				continue // ignore old timeout
+			}
+			log.Debug("Node data timeout", "peer", req.peer.id, "items", len(req.items))
+			s.requestDone(req)
+			s.retry = append(s.retry, req.items...)
+			for _, hash := range req.items {
+				req.peer.MarkLacking(hash)
+			}
+			req.peer.SetNodeDataIdle(0)
+
+		case <-newPeer:
+			// assign new tasks
+		case <-s.quit:
+			return errCancelStateFetch // aborted through s.cancel()
+		case <-s.d.cancelCh:
+			return errCancelStateFetch // downloader was canceled
+		}
+	}
+}
+
+func (s *stateSync) requestDone(req *stateReq) {
+	delete(s.fetching, req.peer.id)
+	close(req.cancelTimeout)
+}
+
+func (s *stateSync) assignTasks() {
+	peers, _ := s.d.peers.NodeDataIdlePeers()
+	for len(peers) > 0 && (s.sched.QueueLen() > 0 || len(s.retry) > 0) {
+		if peer := peers[0]; s.fetching[peer.id] == nil {
+			rtt := s.d.requestRTT()
+			items := s.popTasks(peer.NodeDataCapacity(rtt))
+			s.fetching[peer.id] = s.spawnFetch(peer, items, rtt)
+		}
+		peers = peers[1:]
+	}
+}
+
+func (s *stateSync) popTasks(n int) []common.Hash {
+	items := make([]common.Hash, 0, n)
+	// add from retry list
+	for ; n >= 0 && len(s.retry) > 0; n-- {
+		items = append(items, s.retry[len(s.retry)-1])
+		s.retry = s.retry[:len(s.retry)-1]
+	}
```

Comment:
Note however that maintaining such a list and avoiding reassigning the same hash over and over again will cause the next line to potentially OOM the node, because there would be no memory cap on the total size of the pending state retrievals.

The original code moved trie requests from the scheduler.Missing to an intermediate pool (similar to your `retry` slice, but one that contained all pending ones), and only requested new missing ones if it fit into the pool. You could do a similar thing by maintaining a counter with the total pending requests, and do:

```go
return append(items, s.sched.Missing(min(n, maxPend - curPending))...)
```

That would ensure that we don't grow the number of outstanding requests unboundedly.